docs/design/specs/Ecma-335-Augments.md

```diff
@@ -366,3 +366,72 @@ In addition to the guarantees that apply to all type initializers, the CLI shall
 2. A module initializer shall run exactly once for any given module unless explicitly called by user code.
 
 3. No method other than those called directly or indirectly from the module initializer will be able to access the types, methods, or data in a module before its initializer completes execution.
+
+## Static Interface Methods
+
+This WIP section is intended to specify ECMA standard changes implied by the static interface methods. For the reference ECMA spec, I have used
+
+https://www.ecma-international.org/publications-and-standards/standards/ecma-335/
+
+(6th edition, June 2012), please let me know if I should look at a different version. It also kind of sucks that we cannot reference the ECMA spec
+dynamically via some hyperlinks - I assume there's no online version of the spec that would support referring to the individual sections and such.
+
+Based on initial audit I believe changes potentially need to be made at least to the following sections:
+
+* Virtual methods (I.8.4.4, page 25)
+* Explicit method overrides (II.9.10, page 136)
+* Constraints on generic parameters (II.9.11, page 137)
+* Introducing and overriding virtual methods (II.10.3, page 147)
+* Semantics of interfaces (II.12, page 157)
+* Interface implementation examples (II.12.2.1, page 159)
+* Defining, referencing and calling methods (II.15, page 177)
+* constrained. - (prefix) invoke a member on a value of a variable type (III.2.1, page 316)
+* call - call a method (III.3.19, page 342)
+
+We could use the "interface implementation examples" section to outline some positive and negative examples that can be also used as a basis for
+testing.
+
+Some aspects that are unclear to me now:
+
+* Especially in light of the default interface work, would it make sense to allow "overriding" (implementing) a static interface method
+  in a derived interface? If we don't plan to allow this, is it just to scope the work or does some fundamental issue prevent it?
+* What is the reason why we only allow "abstract" static interface method declarations, not "virtual" in the sense that the interface
+  would provide a default implementation that a class implementing the interface could optionally override?
+
+### I.8.4.4, Virtual Methods
+
+(Add second paragraph)
+
+Static interface methods may be marked as virtual. Valid object types implementing such interfaces shall provide implementations
+for these methods by means of Method Implementations (II.15.1.4). Polymorphic behavior of calls to these methods is facilitated
+by the constrained. call IL instruction where the constrained. prefix specifies the type to use for lookup of the static interface
+method.
+
+### II.9.10, Explicit Method Overrides
+
+(It is unclear to me what is the expected general behavior of generic static virtual methods; it should probably be stated here).
+
+### II.9.11, Constrains on Generic Parameters
```

Comment:
There are two ways to handle this.

1. Require that even in the presence of an abstract type, the static interface methods must be fully implemented. (This would be a change to II.12.2, "If the current class is not abstract and..." would need qualification to specify that static interface methods must be implemented. This would cover the instantiation over class case, but there is another detail, which is that instantiated over the interface which defines the constraint which also need to be disabled. Thus I prefer option 2.
OR
2. We adjust the rules for constraints. Section II.9.11 is adjusted so that instead of specifying that a constraint requires an "assignable-to" relationship, that the relationship is "assignable-to" and *implements-static-interface-methods-of*. Where *implements-static-interface-methods-of* requires that the instantiating type satisfy the interface static methods.

If we go down option 2, we can also have the question of do we implement a type load check at all. My preference is to do both, where we have a check for concrete types at type load time, and the constraint check is also updated to cover the abstract class scenario as well as the interface scenario.

Reply:
I have adjusted section II.9.11 according to David's "2nd option" in 2nd commit.

Reply:
Thanks for clarifying, I'm closing this conversation for now as I believe we've reached a resolution. Please reopen it if you believe there's more to discuss.